src/geo/ui/legends/layer-legends-view.js

```diff
@@ -29,29 +31,67 @@ var LayerLegendsView = Backbone.View.extend({
   },
 
   render: function () {
-    var showLegends = this.settingsModel.get('showLegends');
-    var showLayerSelector = this.settingsModel.get('showLayerSelector');
-    var shouldVisible = showLayerSelector || this.model.legends.hasAnyLegend() && showLegends;
+    var showLegends = this._shouldLegendsBeVisible();
+    var showLayerSelector = this._shouldLayerSelectorBeVisible();
+    var shouldVisible = this._shouldLayerLegendsBeVisible();
+
+    this.$el.html(
+      template({
+        shouldVisible: shouldVisible,
+        layerName: this.model.getName(),
+        isLayerVisible: this._isLayerVisible(),
+        showLegends: showLegends,
+        showLayerSelector: showLayerSelector
+      })
+    );
 
     if (shouldVisible) {
-      this.$el.html(
-        template({
-          layerName: this.model.getName(),
-          isLayerVisible: this._isLayerVisible(),
-          showLegends: showLegends,
-          showLayerSelector: showLayerSelector
-        })
-      );
-
       this._renderLegends();
-    } else {
-      this.$el.empty();
     }
 
     this.tryContainerVisibility();
     return this;
   },
 
+  _shouldLegendsBeVisible: function () {
+    var showLegends = this.settingsModel.get('showLegends');
+    return showLegends && this._isLayerVisible();
+  },
+
+  _shouldLayerLegendsBeVisible: function () {
+    var isEmbed = this._isEmbed;
+    var showLayerSelector = this.settingsModel.get('showLayerSelector');
+    var showLegends = this.settingsModel.get('showLegends');
+    var isSettingsView = !!this.settingsModel.get('settingsView');
```

Comment:
Same here... CartoDB.js shouldn't know anything about "settings"... Are we initialising this anyway?